Memory analyzer to take memory snapshots [blocks: #2649, #4438]

danpoe opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

mmuesly committed:
Add a highlevel api to test types on exprt and typet

The CBMC code base normaly checks the id of an irept subtype
against some magic id. But for some tests, it is necessary to check against
more than one id. So this functions are use do the checks.

This should further support code readability as a combination of different ids
gets a name.

mmuesly committed:
Add the actual core dump analyzer

This code takes a static symbol, zero initalizes the typet of the symbol
and then fills it up with values if possible.
Following the declared structure in the typet, the analyzer
queries the gdb api for values of the attributes of typet.
It the symbol is a primtive type, the analyzer directly
queries for the value of the symbol.

mmuesly committed:
Add regression tests for the memory analyzer

These test have been used as driving example for the memory analyzer.
They test basic functionality and handling of cycles in structs.

mmuesly committed:
Add Makefiles enabeling memory-analyzer and tests

danpoe committed:
Update gdb api to return more information about pointed-to objects

This update the gdb api to return more information about pointers (via the
method get_memory() which returns an object of type pointer_valuet describing
the pointer and pointed-to data). Unit tests for the new functionality are
included.

danpoe committed:
Update existing memory analyzer regression tests

This adds an updated chain.sh file, and disables the existing regression tests.
They currently fail due to two issues: goto code to c conversion currently
ignores array sizes (i.e., even int arrays of known sizes are output as int[]),
and void pointers are not handled (i.e., analyze_symbol() may try to zero
initialize an object of type void, yielding an invariant violation).

danpoe committed:
Remove obsolete functions is_pointer(), is_struct(), etc. from std_expr.h

A previous commit introduced e.g. function is_pointer() to check type.id() ==
ID_pointer. These have become obsolete as these expressions are now used
directly.

danpoe committed:
Update memory analyzer driver and options parsing

danpoe committed:
Update symbol_analyzert

This updates the symbol analyzer which is used by the memory analyzer. The class
is refactored and updated to use the new gdb api.

danpoe committed:
Add additional regression tests for the memory analyzer

xbauch commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+
+/// \file
+/// This file contains functions, that should support test for underlying
+/// c types, in cases where this is required for anlysis purpose.
```
analysis

xbauch commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+  return type.id() == ID_c_bool;
+}
+
+/// This function checks, whether the type is has been some kind of integer
```
is has been -> has been

xbauch commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+/// \param c_enum the enum type memeber_name is supposed to be part of.
+/// \return value a constant, that could be assigned as value for an expression
+/// with type c_enum.
+constant_exprt convert_memeber_name_to_enum_value(
```
member

xbauch commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+/// This function creates a constant representing the
+/// bitvector encoded integer value of a string in the enum.
+/// \param member_name is a string that should be in the enum.
+/// \param c_enum the enum type memeber_name is supposed to be part of.
```
\p member_name

xbauch commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+  {
+    if(id2string(enum_value.get_identifier()) == member_name)
+    {
+      mp_integer int_value = string2integer(id2string(enum_value.get_value()));
```
Could this the call the safe version returning `optionalt`?

hannes-steffenhagen-diffblue replied:
This should just be `numeric_cast<mp_integer>(enum_value.get_value())` I think

This is just trying to convert constants between different types -  I think we might have a function for that already, but I'm not sure

xbauch commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+/// bitvector encoded integer value of a string in the enum.
+/// \param member_name is a string that should be in the enum.
+/// \param c_enum the enum type memeber_name is supposed to be part of.
+/// \return value a constant, that could be assigned as value for an expression
```
This sounds strange.

hannes-steffenhagen-diffblue commented on regression/memory-analyzer/arrays/arrays.h:
```diff
@@ -0,0 +1,32 @@
+struct a_typet
+{
+  int config[10];
+  bool initalized;
```
:pick: initialized

xbauch replied:
Ok.

xbauch commented on src/memory-analyzer/memory_analyzer_parse_options.cpp:
```diff
@@ -0,0 +1,188 @@
+    << " --breakpoint <breakpoint>    analyze from breakpoint\n"
+    << " --symbols <symbol-list>      list of symbols to analyze\n"
+    << " --symtab-snapshot            output snapshot as symbol table\n"
+    << " --output-file <file>         write snapshot to file\n"
```
Could we add a `--json-ui` here, since it's somewhat useful for `goto-harness`?

danpoe replied:
Currently we have the option `--symtab-snapshot` which outputs the memory snapshot in json format.

xbauch replied:
```
$ ../../../build/bin/memory-analyzer --symtab-snapshot --breakpoint checkpoint --symbols array main_exe

Symbols:

Symbol......: __CPROVER_size_t
Pretty name.: __CPROVER_size_t
Module......: main
Base name...: __CPROVER_size_t
Mode........: C
Type........: __CPROVER_size_t
Value.......: 
Flags.......: thread_local file_local type macro
Location....: file <built-in-additions> line 1

Symbol......: array
Pretty name.: array
Module......: main
Base name...: array
Mode........: C
Type........: signed int [3l]
Value.......: { 1, 4, 3 }
Flags.......: lvalue static_lifetime
Location....: file main.c line 5
```
Am I using it wrong?

danpoe replied:
Oh, you're right. `--json-ui` is indeed needed and missing from the help section.

hannes-steffenhagen-diffblue commented on src/ansi-c/c_typecast.cpp:
```diff
@@ -55,14 +55,14 @@ bool c_implicit_typecast_arithmetic(
 }
 
-bool is_void_pointer(const typet &type)
+bool has_a_void_pointer(const typet &type)
```
Unrelated rename and whitespace change?

Commit-Id: 7bcfc117d15d9aff20398285eae1741c6d94c7e0

xbauch commented:
https://github.com/danpoe/cbmc/pull/7 .. here is a documentation PR for this tool.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cb7ce3e86a0ffbf956ca837899d3fbc77e745162).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105120198

martin-cs commented:
@danpoe : I got a request to review this, apparently from you, because I was a code owner?  I am happy to review if you want but this seems like a mistake and that it might not actually help you get it merged.

hannes-steffenhagen-diffblue commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+// Copyright 2018 Author: Malte Mues <mail.mues@gmail.com>
```
Doesn't comply with our file headers

hannes-steffenhagen-diffblue commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+/// \return value a constant, that could be assigned as value for an expression
+/// with type c_enum.
+constant_exprt convert_memeber_name_to_enum_value(
+  const std::string &member_name,
```
`irep_idt`

xbauch replied:
Fixed.

hannes-steffenhagen-diffblue commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+/// \param bool_value: A string that is compared to "true" ignoring case.
+/// \param type: The type, the resulting constant is supposed to have.
+/// \return a constant of type \param type with either 0 or 1 as value.
+constant_exprt from_c_boolean_value(std::string bool_value, const typet &type)
```
`irep_idt`? Also, this'd probably be better as two functions, one `string -> bool` and one `bool -> expr` - and I think both of those exist already.

xbauch replied:
Couldn't find them: wrote my own.

hannes-steffenhagen-diffblue commented on src/util/std_types.h:
```diff
@@ -25,6 +25,13 @@ Author: Daniel Kroening, kroening@kroening.com
 
+/// This method tests,
+/// if the given typet is a constant
+inline bool is_constant(const typet &type)
```
I'd be rather surprised if this was ever true. `ID_constant` usually marks expression literals (literal expressions? you know what I mean)

tautschnig replied:
Yes, this one is certainly wrong. If it's about a type having a `const` qualifier, then it needs to return `type.get_bool(ID_C_constant);`

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+{
+}
+
+void symbol_analyzert::analyse_symbol(const std::string &symbol_name)
```
`irep_idt`

xbauch commented:
Regarding the `pointer_to_struct` regression test:
```
$ memory-analyzer --symtab-snapshot --json-ui --breakpoint checkpoint --symbols p main_exe
```
violates an invariant (`gdb_api.cpp:142`) even though running it without `--symtab-snapshot --json-ui` works.
Also running
```
$ memory-analyzer --symtab-snapshot --json-ui --breakpoint checkpoint --symbols st,p main_exe
```
works.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+// Copyrigth 2018 Author: Malte Mues <mail.mues@gmail.com>
```
The commit message in 079f28a6ce76c59fd8632aa7fa37ee0a5aa6b389 does give a short summary of the algorithm, instead it should explain what this is supposed to be for (or rather: It's not like the summary needs to go away, but the main purpose of the commit message should be giving context for the commit)

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+#include <util/std_types.h>
+#include <util/string_constant.h>
+
+symbol_analyzert::symbol_analyzert(
```
`symbol_analyzer` is perhaps an overly generic name for what this is doing?

xbauch replied:
`gdb_value_extractort`?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+void symbol_analyzert::analyse_symbol(const std::string &symbol_name)
+{
+  const symbolt symbol = ns.lookup(symbol_name);
+  const symbol_exprt symbol_expr(symbol.name, symbol.type);
```
`symbol.symbol_expr()`

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  }
+  else if(is_pointer(symbol.type))
+  {
+    const std::string memory_location = gdb_api.get_memory(symbol_name);
```
Why is this bit inlined whereas some of the other handlers are put in separate functions?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  {
+    const std::string memory_location = gdb_api.get_memory(symbol_name);
+
+    if(memory_location != "0x0")
```
This appears to assume that any pointer that isn't a nullpointer is valid, which isn't generally the case.

xbauch replied:
Well I don't really know how to fix it but here it's only a sort of optimisation. Afterwards, e.g. in `get_non_char_pointer_value` we search our own memory map and if this particular `memory_location` is not present we return a dummy. So it is safe in the sense that we do not try to a value from an invalid memory address.

xbauch commented:
This is probably a terrible hack, but I would like to have the `address_of_exprt` symbol expressions to have the right type:
```
modified   src/memory-analyzer/analyze_symbol.cpp
@@ -232,7 +232,9 @@ exprt symbol_analyzert::get_non_char_pointer_value(
   }
   else
   {
-    return it->second;
+    auto result = it->second;
+    result.type() = expr.type();
+    return result;
   }
 }
 
@@ -261,6 +263,7 @@ exprt symbol_analyzert::get_pointer_value(
     {
       const exprt target_expr =
         get_non_char_pointer_value(expr, memory_location, location);
+      INVARIANT(target_expr.type()==zero_expr.type(), "target has wrong type");
 
       if(target_expr.id() == ID_nil)
       {
```

tautschnig commented:
@xbauch what about adding a typecast? `auto result = typecast_exprt::conditional_cast(it->second, expr.type());`

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+    }
+    else
+    {
+      const exprt null_ptr = zero_initializer(symbol.type, symbol.location, ns);
```
Why not just `from_integer`?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+    {
+      warning() << "Couldn't access memory location for " << symbol_name << "\n"
+                << "continue execution anyhow."
+                << "You might want to check results for correctness!"
```
At the very least the last line here seems to be unnecessary

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  const namespacet &ns)
+{
+  const std::regex single_char_pattern(
+    "(?:(?:(?:\\\\)?([0-9]{3}))|([\\\\]{2}|\\S))");
```
This regex could really use an accompanying explanation.

chrisr-diffblue replied:
Definietly - a little example of the input at the resulting match would be really helpful

xbauch commented:
@tautschnig Thanks! Typecast works as well (and doesn't actually modify the stored expression). In the above I was off by one `subtype`. The fix should look like this:
```
modified   src/memory-analyzer/analyze_symbol.cpp
@@ -232,7 +232,7 @@ exprt symbol_analyzert::get_non_char_pointer_value(
   }
   else
   {
-    return it->second;
+    return typecast_exprt::conditional_cast(it->second, expr.type().subtype());
   }
 }
 
@@ -268,7 +268,9 @@ exprt symbol_analyzert::get_pointer_value(
       }
       else
       {
-        return address_of_exprt(target_expr);
+        const auto result_expr = address_of_exprt(target_expr);
+        CHECK_RETURN(result_expr.type()==zero_expr.type());
+        return result_expr;
       }
     }
   }
```

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  std::vector<std::string> array_content;
+  size_t content_size = type.get_width();
+  while(regex_search(values, result, single_char_pattern))
+  {
```
Should have an `INVARIANT(result.size() == 3,...)` along with an explanation for why it holds?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+    if(!result[1].str().empty())
+    {
+      // Char are octal encoded in gdb output.
+      mp_integer int_value = string2integer(result[1], OCTAL_SYSTEM);
```
Could also use `stoi` here, not sure if it'd be better

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+    else
+    {
+      throw symbol_analysis_exceptiont(
+        "It is not supposed that non group match");
```
I do not know what that error message means. Exceptions should have messages that are meaningful to end users, this isn't. If you think this can't happen at all this should be an invariant anyway.

chrisr-diffblue replied:
Agreed - this seems like the wrong way to handle this error condition.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  const std::string &memory_location,
+  const source_locationt &location)
+{
+  const std::regex octal_encoding_pattern("(\\\\[0-9]{3})");
```
Why do the octal digits here include 8 and 9?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+code_declt symbol_analyzert::declare_instance(const typet &type)
+{
```
If this is supposed to be for new symbols it should use something like `get_fresh_aux_symbol`

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+
+exprt symbol_analyzert::process_any_pointer_target(
+  const symbol_exprt &symbol,
+  const std::string memory_location,
```
Why is a memory location a string?

xbauch replied:
I wrapped the address string into a struct with a minimal interface.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+
+exprt symbol_analyzert::fill_array_with_values(
+  const std::string &symbol_id,
+  exprt &array,
```
The code below seems to assume this is a specific type of `exprt`, if so the parameter here should reflect that.

Also, this function modifies an incoming `exprt` and then also returns it as a copy, which seems highly unnecessary. I'd recommend making this a `const whatever_exprt&` and only returning the modified copy instead.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+
+exprt symbol_analyzert::fill_expr_with_values(
+  const std::string &symbol_id,
+  exprt expr,
```
Const ref, in the cases where it's modified just make a copy and return that.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+{
+  const std::string memory_location =
+    gdb_api.get_memory(id2string(field_access.get_identifier()));
+  if(memory_location != "0x0")
```
again, seems to assume anything other than nullptr is valid?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+
+exprt symbol_analyzert::fill_struct_with_values(
+  const std::string &symbol_id,
+  exprt &expr,
```
`const`, and if it's meant to be a specific kind of expr (which use of `operands` implies) change type to reflect that

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  const struct_typet structt = to_struct_type(type);
+  for(size_t i = 0; i < expr.operands().size(); ++i)
+  {
+    exprt &operand = expr.operands()[i];
```
avoid mutable ref to sub expr unless necessary

xbauch replied:
I agree but how else should one build a struct expression?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  const typet &type,
+  const source_locationt &location)
+{
+  const struct_typet structt = to_struct_type(type);
```
can be const ref

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  const source_locationt &location)
+{
+  const std::string &pointer_name = id2string(pointer_symbol.get_identifier());
+  const std::string &char_casted = "(char *)" + pointer_name;
```
shouldn't be a reference

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  const symbol_exprt &pointer_symbol,
+  const source_locationt &location)
+{
+  const std::string &pointer_name = id2string(pointer_symbol.get_identifier());
```
shouldn't be a reference

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+
+  try
+  {
+    const typet &new_char_pointer = pointer_type(char_type());
```
shouldn't be a reference

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+    // but doesn't corupt gdbs process state. So it's save to continue.
+    // Anyhow, you should inspect the result and value of field_access
+    // manually before using the state for verifcation.
+    // Field access stays zero initalized.
```
I don't quite understand this comment?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -0,0 +1,114 @@
+// Copyright 2018 Author: Malte Mues
+#ifdef __linux__
```
FWIW instead of these ifdefs all over the place we should really just exclude the entire memory analyzer from the build when we're not building on linux (+ maybe OSX? Does this work on OSX?)

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -0,0 +1,114 @@
+  std::string get_code();
+
+private:
+  gdb_apit &gdb_api;
```
Would it be alright to change this so the symbol analyzer owns the GDB API, or will we generally need to do something with it after the analyzer finishes?

xbauch replied:
We do a few things before but nothing that couldn't be encapsulated.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -0,0 +1,114 @@
+  void process_outstanding_assignments();
+};
+
+class symbol_analysis_exceptiont : public std::exception
```
Should inherit from `cprover_exception_baset`

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -0,0 +1,114 @@
+  std::string error;
+};
+
+class place_holder_exprt : public exprt
```
This needs documentation

hannes-steffenhagen-diffblue commented on src/memory-analyzer/memory_analyzer_main.cpp:
```diff
@@ -0,0 +1,16 @@
+// Copyright 2018 Author: Malte Mues <mail.mues@gmail.com>
```
All of these files should have their headers changed to our format

xbauch replied:
Done, hopefully everywhere.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/memory_analyzer_main.cpp:
```diff
@@ -0,0 +1,16 @@
+#else
+int main()
+{
+  throw "only supported on Linux platforms currently\n";
```
again, no reason to build this thing at all on non-linux platforms, just change the Makefile/CMakeLists accordingly

hannes-steffenhagen-diffblue commented on src/memory-analyzer/memory_analyzer_parse_options.cpp:
```diff
@@ -0,0 +1,102 @@
+  }
+  else
+  {
+    error() << "It is required to provide the symbols flag in order to make "
```
just throw an exception

hannes-steffenhagen-diffblue commented on src/memory-analyzer/memory_analyzer_parse_options.cpp:
```diff
@@ -0,0 +1,102 @@
+  }
+  else
+  {
+    error() << "Either the 'core-file' or 'breakpoint; flag must be provided\n"
```
just throw an exception

hannes-steffenhagen-diffblue commented on src/memory-analyzer/memory_analyzer_parse_options.cpp:
```diff
@@ -0,0 +1,102 @@
+
+  if(cmdline.args.size() != 1)
+  {
+    error() << "Please provide one binary with symbols to process" << eom;
```
just throw an exception

hannes-steffenhagen-diffblue reviewed (commented):
Comments for up to and including 079f28a6ce76c59fd8632aa7fa37ee0a5aa6b389

hannes-steffenhagen-diffblue commented on src/CMakeLists.txt:
```diff
@@ -93,6 +93,8 @@ add_subdirectory(json)
 add_subdirectory(json-symtab-language)
 add_subdirectory(langapi)
 add_subdirectory(linking)
+add_subdirectory(memory-analyzer)
```
Make this dependent on building on linux

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -181,6 +182,37 @@ array_exprt create_char_array_from_string(
 
+code_declt
+symbol_analyzert::declare_instance(const std::string &prefix, const typet &type)
+{
```
`get_fresh_aux_symbol` or `allocate_objectst` instead?

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -235,6 +235,7 @@ exprt symbol_analyzert::declare_and_initalize_char_ptr(
 
     code_declt target_object =
       declare_instance(id2string(symbol.get_identifier()), init.type());
+
```
5dc87d254166f30c257e3bc6a0cdf465a02599c5 only adds a newline, which is not what the commit message says?

hannes-steffenhagen-diffblue commented on src/Makefile:
```diff
@@ -18,7 +18,6 @@ DIRS = analyses \
        langapi \
        linking \
        memory-analyzer \
-       memory-models \
```
80debb1667f7fdf1c3594bb2572d45e8334c2e81 has this message:

```
To make private methos avilable to unit tests, it is a common practice to redefine the private keyword.
While GCC doesn't carea bout this, older clang versions don't support this behavior and issue a warning.
As all warnings are converted into errors, this waring breaks the build. To allow the intended behavior
during testing, the -Wno-key-macro flag is added.
```

The commit doesn't do that, and it absolutely shouldn't even if it did.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/gdb_api.cpp:
```diff
@@ -358,32 +360,36 @@ std::string gdb_apit::eval_expr(const std::string &expr)
-      "` is not a memory address or has unrecognised format");
+    const std::size_t len = string.length();
+
+    INVARIANT(len >= 4, "");
```
why do these invariants not have messages?

xbauch replied:
Filled.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/memory_analyzer_parse_options.cpp:
```diff
@@ -1,102 +1,188 @@
 
-#ifdef __linux__
+// clang-format off
+#if defined(__linux__) || \
```
Again, I think it'd just be better to conditionally enable/disable the build entirely

hannes-steffenhagen-diffblue commented on src/memory-analyzer/memory_analyzer_parse_options.cpp:
```diff
@@ -1,102 +1,188 @@
+  if(!(core_file ^ breakpoint))
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "need to provide either option --core-file or option --breakpoint", "");
```
I think it is slightly unclear from the message that the `--core-file` and `--breakpoint` options are mutually exclusive - maybe two separate error messages for when neither or both are set?

xbauch replied:
Ok.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/memory_analyzer_parse_options.cpp:
```diff
@@ -1,102 +1,188 @@
-    std::vector<std::string> result;
-    split_string(symbol_list, ',', result, false, false);
+    throw deserialization_exceptiont(
+      "cannot read goto binary `" + binary + "`");
```
:pick: :pick: :pick: The usual format for quoting we use is \` ', not \`\`

xbauch replied:
Ok.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -23,483 +41,371 @@
+    const pointer_valuet &value = gdb_api.get_memory(c_expr);
+    CHECK_RETURN(value.pointee.empty() || (value.pointee == id));
+
+    values.insert(std::make_pair(value.address, symbol_expr));
```
Could just use `{value.address, symbol_expr}`

xbauch replied:
Ok.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -23,483 +41,371 @@
-  else if(is_pointer(symbol.type))
+  catch(gdb_interaction_exceptiont e)
   {
-    const std::string memory_location = gdb_api.get_memory(symbol_name);
```
I think 9b06412843664d0ab5bd9504260194498a4dd76f could use a bit more verbose commit message - I can't quite tell what's going on here

hannes-steffenhagen-diffblue reviewed (commented):
I'm not quite comfortable with approving this myself as I can't say I completely understand what the parts here are doing

hannes-steffenhagen-diffblue commented:
Also, I'm not sure if it's the best approach but it might be worth considering backporting some of the fixes that were later made to the earlier commits

tautschnig commented on src/ansi-c/c_typecast.cpp:
```diff
@@ -212,10 +212,11 @@ bool check_c_implicit_typecast(
+      else if(
+        has_a_void_pointer(src_type) || // from void to anything
+        has_a_void_pointer(dest_type))  // to void from anything
         return false;
```
Add braces after multi-line condition.

xbauch replied:
Done.

tautschnig commented on src/ansi-c/c_typecast.cpp:
```diff
@@ -212,10 +212,11 @@ bool check_c_implicit_typecast(
       const irept &src_subtype =src_type.subtype();
 
-      if(src_subtype==dest_subtype)
+      if(src_subtype == dest_subtype)
```
Unrelated whitespace change.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  }
+  else
+  {
+    const typet target_type = ns.follow(symbol.type);
```
Please don't do this here, see next comment below.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+      target_type,
+      symbol.location);
+
+    if(is_struct(target_type))
```
Make this one `if(is_struct(ns.follow(symbol.type))` or `if(symbol.type == ID_struct || symbol.type == ID_struct_tag)`

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+
+    if(is_struct(target_type))
+    {
+      const struct_typet structt = to_struct_type(target_type);
```
`to_struct_type(ns.follow(symbol.type));`

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+      for(std::size_t i = 0; i < target_expr.operands().size(); ++i)
+      {
+        auto &member_declaration = structt.components()[i];
+        if(!member_declaration.get_is_padding())
```
`member_declaration.get_is_padding() || member_declaration.type().id() == ID_code`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+code_declt symbol_analyzert::get_pointer_target(
+  const std::string pointer_name,
```
Should this really be a `std::string` and not an `irep_idt`?

xbauch replied:
Since we interface gdb, which doesn't really know about ireps, I don't see much added value in using `irep_idt` in some places. Especially if we run `std::string` operations on them afterwards.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+    return convert_memeber_name_to_enum_value(
+      value, to_c_enum_type(expr.type()));
+  }
+  else if(is_struct(type))
```
Should better be `else if(type.id() == ID_struct || type.id() == ID_struct_tag)`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+exprt symbol_analyzert::fill_struct_with_values(
+  const std::string &symbol_id,
+  exprt &expr,
+  const typet &type,
```
Should be a `struct_typet` to ensure the caller is aware of the type constraints.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  for(size_t i = 0; i < expr.operands().size(); ++i)
+  {
+    exprt &operand = expr.operands()[i];
+    const typet &resolved_type = ns.follow(operand.type());
```
Shouldn't be done here, only resolve were actually needed (i.e., with the is-it-a-struct-type below)

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  {
+    exprt &operand = expr.operands()[i];
+    const typet &resolved_type = ns.follow(operand.type());
+    if(!structt.components()[i].get_is_padding())
```
Also test for `component_type.id() == ID_code`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+          location);
+
+        declaration.operands().resize(2);
+        declaration.op1() = target_expr;
```
This is a recurring pattern - maybe `declare_instance` should just take a second parameter?

tautschnig commented on unit/Makefile:
```diff
@@ -156,7 +156,12 @@ CLEANFILES = $(CATCH_TEST) testing-utils/testing-utils$(LIBEXT)
+=======
+CLEANFILES = $(TESTS)\
+        #Empty last line
+>>>>>>> cf37266... Add Makefiles enabeling memory-analyzer and tests
```
Merge conflict?

tautschnig commented on regression/memory-analyzer/Makefile:
```diff
@@ -0,0 +1,21 @@
+default: clean tests.log
```
Could we also have a CMakeLists.txt, please?

xbauch replied:
Added.

tautschnig commented on regression/memory-analyzer/Makefile:
```diff
@@ -0,0 +1,21 @@
+	find -name '*.out' -execdir $(RM) '{}' \;
+	$(RM) tests.log
+
+test:
```
The regression test directory also needs to be added to the parent directory's Makefile and CMakeLists.txt.

xbauch replied:
Done.

tautschnig commented on unit/Makefile:
```diff
@@ -156,12 +156,7 @@ CLEANFILES = $(CATCH_TEST) testing-utils/testing-utils$(LIBEXT)
-=======
-CLEANFILES = $(TESTS)\
-        #Empty last line
->>>>>>> cf37266... Add Makefiles enabeling memory-analyzer and tests
```
Squash with the previous commit.

tautschnig commented on src/memory-analyzer/Makefile:
```diff
@@ -13,7 +13,7 @@ LIBS = \
 	../big-int/big-int.a \
 	../langapi/langapi.a
-	
+
```
Unnecessary change

xbauch commented:
@tautschnig and @hannes-steffenhagen-diffblue I have addressed the comments that were still visible in the last commit of https://github.com/danpoe/cbmc/pull/7. The first two commits there add documentation and then @hannes-steffenhagen-diffblue changed the way `memory-analyzer` is included in the build. I did not address things that would require my changing the previous commits (white spaces, commit messages, etc.).

chrisr-diffblue commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+#include <string>
+
+/// This function checks, whether this has been a char type in the c program.
+inline bool is_c_char(const typet &type)
```
I wonder if the naming of these new functions should be `is_blah_type` ? The thing being check is a *type*, not a value.

chrisr-diffblue commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+  const irep_idt &c_type = type.get(ID_C_c_type);
+  return is_signed_or_unsigned_bitvector(type) &&
+         (c_type == ID_char || c_type == ID_unsigned_char ||
+          c_type == ID_signed_char);
```
Should this also handle all the `wchar_t` types as well?

chrisr-diffblue commented on src/util/c_types_util.h:
```diff
@@ -0,0 +1,109 @@
+/// type in the c program.
+/// It considers the signed and unsigned verison of
+/// int, short, long and long long as integer types in c.
+inline bool is_c_int_derivate(const typet &type)
```
Would a better name be `is_c_integral_type` ?

chrisr-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,476 @@
+  else
+  {
+    throw symbol_analysis_exceptiont(
+      "unexpected thing:  " + expr.pretty() + "\nexception end\n");
```
Probably should have a better exception message here, or possibly an invariant

chrisr-diffblue commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -0,0 +1,114 @@
+class exprt;
+class source_locationt;
+
+class symbol_analyzert : public messaget
```
Probably shouldn't inherit from `messaget`? `symbol_analyzert` isn't *a* logger, it *has* a logger.

xbauch replied:
@chrisr-diffblue I addressed the comments which weren't fixed already in https://github.com/danpoe/cbmc/pull/7.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: baf3624da97f607aac96fa214e82e04416f2c90e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108877818
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

danpoe commented:
Ok, so we've addressed all the comments on this PR. @mmuesly's commits from #2649 were included as is, so issues in them are only fixed in later commits.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9443c9322f4ae163441a5bd28f0ca51471cf5e6a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/109310640
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: bcd07a0cb13f814f6124ca20048bf146c34d7fa5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/109314125
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

tautschnig commented on doc/cprover-manual/memory-analyzer.md:
````diff
@@ -0,0 +1,78 @@
+(having `goto-gcc` on the `PATH`):
+
+```sh
+$ goto-gcc -g -std=c11 input_program.c -o input_program_exe
````
Is `-std=c11` really important here? (Examples should always be limited to the required minimum so as not to distract from the key point.)

xbauch replied:
Well, it's not necessary for this example, so removed. Also I'll be addressing it the comments here: https://github.com/danpoe/cbmc/pull/9.

tautschnig commented on doc/cprover-manual/memory-analyzer.md:
````diff
@@ -0,0 +1,78 @@
+```
+
+Calling `goto-gcc` instead of simply compiling with `gcc` produces an ELF binary
+with a goto section that contains the goto model (goto program + symbol table).
````
Hmm, do we not document this anywhere else? If we don't, that's a problem, else we should have a cross reference here.

xbauch replied:
There is a (somewhat terse) description in `doc/cprover-manual/goto-cc-variants.md`. Link included.

tautschnig replied:
Even if terse, it seems like the right place for such documentation. So please do add a reference to that bit of documentation, in addition to what you are documenting about it here.

tautschnig commented on doc/cprover-manual/memory-analyzer.md:
```diff
@@ -0,0 +1,78 @@
+with a goto section that contains the goto model (goto program + symbol table).
+
+The user can choose to either run GDB from a core-file `--core-file cf` or to a
+break-point `--breakpoint bp` (and only one of those). The tool also expects a
```
So you're telling me I have a choice, but I don't yet know what the overall workflow is - and thus I cannot make an informed decision. It would be good to include a brief paragraph that sketches the steps involved at very high level (compile the program, run it without GDB and obtain a core and then load the core in GDB _or_ run it within GDB directly, and then ...).

xbauch replied:
Explained and expanded.

tautschnig commented on doc/cprover-manual/memory-analyzer.md:
````diff
@@ -0,0 +1,78 @@
+`checkpoint`. We compile the program with
+
+```sh
+$ goto-gcc -g -std=c11 main.c -o main_exe
````
See above re `-std=c11`

xbauch replied:
Removed.

tautschnig commented on doc/cprover-manual/memory-analyzer.md:
````diff
@@ -0,0 +1,78 @@
+```sh
+$ memory-analyzer --symtab-snapshot --json-ui \
+--breakpoint checkpoint --symbols array main_exe > snapshot.json
+```
````
For the last two paragraphs: how to interpret the output generated (whether JSON or plain text)? I think this needs to be explained here.

xbauch replied:
Explained.

tautschnig commented on regression/Makefile:
```diff
@@ -27,6 +27,10 @@ DIRS = cbmc \
 
+ifeq ($(WITH_MEMORY_ANALYZER),1)
+  DIRS += memory-analyzer
+endif
```
Do we really need to make this conditional? Can't we just make this dependent on the OS/`BUILD_ENV_`? At least I'd much prefer to keep things as simple as possible.

hannes-steffenhagen-diffblue replied:
@tautschnig We could enable the flag by default on Linux if you'd prefer, but the memory analyzer (and its tests!) don't actually work on some platforms (including some linux configurations), so I think just having a single flag saying that you want it is already the "simple" version.

tautschnig replied:
What is it that keeps is from working across all Linux configurations? I'm worried that anything that is optional will die at some point, and even when it doesn't people might not give it a whole lot of testing in their local builds. So, if we can, we should just enable it by default.

hannes-steffenhagen-diffblue replied:
It's slightly exotic, but some Linux systems have `ptrace` disabled or heavily restricted: https://www.kernel.org/doc/Documentation/security/Yama.txt - I believe we had at least one of the CI environments here configured this way, so `memory-analyzer` actually didn't work in CI for a while. Slightly less exotic is the possibility of people not having `gdb` installed.

Again, I don't mind turning it on by default on Linux as it's going to work _most_ of the time, but I'd also like to keep a user controlled option to turn it off if they don't want it, or forcing it to turn on even if our platform detection logic thinks it wouldn't work.

tautschnig replied:
Ok, so maybe turning it on by default is the best option. And, if possible, I'd prefer to build whenever possible, and only restrict testing.

xbauch replied:
So always build on linux, and test there by default. Unchange on other platforms. Something like this: https://github.com/danpoe/cbmc/pull/9/commits/d22cb3c26931f4c2fa5108bbef2e027f282c2d88?

tautschnig commented on regression/memory-analyzer/arrays/test.desc:
```diff
@@ -0,0 +1,24 @@
+FUTURE
```
Still `FUTURE`? If yes, please add comments (at the end of the file, after two `--` lines) to explain what's missing at the moment.

xbauch replied:
Explanation added (we cannot extract individual `char` values, only strings). The regression tests changes are in https://github.com/danpoe/cbmc/pull/9/commits/ed27ee66829bc8f0a15161e7d9b6f34dc91e47af.

tautschnig commented on regression/memory-analyzer/chain.sh:
```diff
@@ -0,0 +1,12 @@
+args=${*:3:$#-3}
+
+$goto_gcc -g -std=c11 -o "${name}.exe" "${name}.c"
+$memory_analyzer $args "${name}.exe"
```
I'm not sure I'd use `.exe` when we're explicitly Unix-only. Probably `.bin`. But also not a big deal, might not be worth the trouble changing this.

xbauch replied:
We use `.gb` in `goto-harness`. Changed.

tautschnig commented on regression/memory-analyzer/cycles/test.desc:
```diff
@@ -0,0 +1,16 @@
+FUTURE
```
As above.

xbauch replied:
This one actually works. Changed.

tautschnig commented on regression/memory-analyzer/pointer_08/test.desc:
```diff
@@ -0,0 +1,5 @@
+KNOWNBUG
```
Same comment as `FUTURE`.

xbauch replied:
The requested symbol has type `void*`. Explanation added.

tautschnig commented on regression/memory-analyzer/primitive_types/primitive_types.c:
```diff
@@ -0,0 +1,42 @@
+/// Pointer references are tested and ensured, that for example f and f_1 are
+/// pointing to the same int value location after running memory-analyzer.
+
+#include "primitive_types.h"
```
The technical term is "pod" (plain old datatype) I'd say. Which just sounds less dismissive.

xbauch replied:
Renamed.

tautschnig commented on regression/memory-analyzer/primitive_types/test.desc:
```diff
@@ -0,0 +1,5 @@
+FUTURE
```
As above.

xbauch replied:
This one also works.

tautschnig commented on regression/memory-analyzer/strings_01/test.desc:
```diff
@@ -0,0 +1,8 @@
+KNOWNBUG
```
As above.

xbauch replied:
Also works.

tautschnig commented on regression/memory-analyzer/structs_03/test.desc:
```diff
@@ -0,0 +1,8 @@
+KNOWNBUG
```
As above.

xbauch replied:
Works.

tautschnig commented on regression/memory-analyzer/structs_04/test.desc:
```diff
@@ -0,0 +1,6 @@
+KNOWNBUG
```
As above.

xbauch replied:
Explained (single char extraction unsupported for now).

tautschnig commented on regression/memory-analyzer/void_pointer/test.desc:
```diff
@@ -0,0 +1,5 @@
+KNOWNBUG
```
As above.

xbauch replied:
Explained (void pointer).

tautschnig commented on src/CMakeLists.txt:
```diff
@@ -104,3 +104,7 @@ add_subdirectory(goto-instrument)
+
+if(WITH_MEMORY_ANALYZER)
+add_subdirectory(memory-analyzer)
+endif()
```
Like the Makefile case, I'd prefer to link this to OS/build environment rather than introducing a new option.

tautschnig commented on src/Makefile:
```diff
@@ -17,19 +17,27 @@ DIRS = analyses \
+
+ifeq ($(WITH_MEMORY_ANALYZER),1)
+  ALLDEPS += memory-analyzer.dir
+endif
```
This introduction of `ALLDEPS` actually isn't even necessary as you can safely do
```
ifeq ($(WITH_MEMORY_ANALYZER),1)
all: memory-analyzer.dir
endif
```

xbauch replied:
Changed. This and the remaining changes (in makefiles and the memory-analyzer command line options) are in https://github.com/danpoe/cbmc/pull/9/commits/9af754fd5c3cd6ead3d7b0b1a45327b07c45f8d9.

tautschnig commented on src/memory-analyzer/Makefile:
```diff
@@ -0,0 +1,37 @@
+SRC = analyze_symbol.cpp\
```
Nit picking: space before `\`

xbauch replied:
Added.

tautschnig commented on src/memory-analyzer/Makefile:
```diff
@@ -0,0 +1,37 @@
+SRC = analyze_symbol.cpp\
+			gdb_api.cpp \
+			memory_analyzer_main.cpp \
+			memory_analyzer_parse_options.cpp
```
Mix of tabs and spaces?

xbauch replied:
I cleaned the whitespaces in this file a bit more.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,375 @@
+    symbol_table(symbol_table),
+    ns(symbol_table),
+    c_converter(ns, expr2c_configurationt::clean_configuration),
+    allocate_objects(ID_C, source_locationt(), "", this->symbol_table)
```
Use `irep_idt{}` instead of `""`, and I don't this the use of `this->` is necessary here.

xbauch replied:
Done for the first part, but no for the second: if we don't explicitly refer to the member `symbol_table` we bind the constant reference the constructor got as its argument. And `allocate_objects` takes a non-const reference.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,375 @@
+}
+
+void gdb_value_extractort::analyze_symbols(
+  const std::vector<std::string> &symbols)
```
I think this should be a `std::vector<irep_idt>` to avoid repeated `std::string` -> `irep_idt` conversion.

xbauch replied:
No problem, but the way we get the vector is by calling `split_string`. I generalised it to take `irep_idt`s as well. Now it seems to be a bit bigger change that originally intended (https://github.com/danpoe/cbmc/pull/9/commits/38b9cb1d1d57ba08589ddb89e787a38ab26dc3f0)

xbauch replied:
It was suggested that `split_string` actually shouldn't work for `irep_idt`s. I changed the fix and now we transform the vector of strings (from `split_string`) to a vector of `irep_idt` before passing it to `analyze_symbols`. Is this solution ok https://github.com/danpoe/cbmc/pull/9/commits/b4555e55e4f71fe88e4c7560c146b30a2cebcc8e?

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,375 @@
+  // record addresses of given symbols
+  for(const auto &id : symbols)
+  {
+    const symbol_exprt symbol_expr(id, typet());
```
I'd suggest to use `symbol_exprt::typeless`. Though, reading a bit further: maybe actually `const symbol_exprt symbol_expr = ns.lookup(id).symbol_expr();`?

xbauch replied:
Opted for the latter. This and the remaining changes within `analyze_symbol.cpp` are in https://github.com/danpoe/cbmc/pull/9/commits/b4555e55e4f71fe88e4c7560c146b30a2cebcc8e.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,375 @@
+    std::string c_expr = c_converter.convert(expr);
+    pointer_valuet value = gdb_api.get_memory(c_expr);
+    CHECK_RETURN(value.string);
+    std::string string = *value.string;
```
This temporary doesn't seem to add much value? Maybe just do `string_constantt init(*value.string);`?

xbauch replied:
Inlined.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,375 @@
+    string_constantt init(string);
+    CHECK_RETURN(to_array_type(init.type()).is_complete());
+
+    symbol_exprt dummy(pointer_typet(init.type(), config.ansi_c.pointer_width));
```
Use `pointer_type` from `util/c_types.{h,cpp}` and it's missing a name.

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,375 @@
+
+    const typet target_type = expr.type().subtype();
+
+    symbol_exprt dummy(expr.type());
```
Missing name (this constructor actually is deprecated).

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,375 @@
+
+    return get_pointer_value(expr, zero_expr, location);
+  }
+  UNREACHABLE;
```
`UNIMPLEMENTED` maybe? There certainly are a number of types that haven't been handled above.

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,375 @@
+{
+  PRECONDITION(zero_expr.id() == ID_struct);
+
+  PRECONDITION(expr.type().id() == ID_struct_tag);
```
Hmm, this can also be called with `expr.type().id() == ID_struct`, although I do think that that should actually never be the case, i.e., the `ID_struct` case should be removed from the call site.

xbauch replied:
I added a precondition to `get_expr_value` and removed the case from its analysis.

tautschnig commented on src/memory-analyzer/memory_analyzer_parse_options.cpp:
```diff
@@ -0,0 +1,183 @@
+    << "\n"
+    << " --core-file <file>           analyze from core file\n"
+    << " --breakpoint <breakpoint>    analyze from breakpoint\n"
+    << " --symbols <symbol-list>      list of symbols to analyze\n"
```
Should be explicit that this is a required option

xbauch replied:
Made explicit.

tautschnig commented on doc/cprover-manual/memory-analyzer.md:
````diff
@@ -0,0 +1,78 @@
+And then we call `memory-analyzer` with:
+
+```sh
+$ memory-analyzer --breakpoint checkpoint --symbols array main_exe
````
Hmm, so GDB never needs to be run directly? The above text made it sound like that would have been necessary.

xbauch replied:
Added.

tautschnig reviewed (commented):
I think my biggest concern are the build-system changes as everything else is very local. See various comments for details.

danpoe commented:
Ok, so @xbauch's PR addressing the latest comments has now been merged, and the PR was rebased on the latest develop.

tautschnig reviewed (approved):
It might be good to apply some squashing, but otherwise looks good from my point of view.

chrisr-diffblue reviewed (approved):
Looks good to me - thanks to everyone involved in creating, debugging, and reviewing this PR - it's been quite an effort!

I would echo Michael's comment about maybe doing some squashing though :-)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c19aa7b88cdb388e0257310331713b8cf51a074c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111901196

tautschnig commented:
I'll leave merging to @danpoe or @xbauch to do some squashing (or consider it too much overhead and just merge straight away).

hannes-steffenhagen-diffblue committed:
Make memory analyzer optional

We were doing this in multiple places, this should make it a bit easier to keep
set properties consistent among different places.

Ideally we could eventually move away from mentioning targets from modules here
on the top level (and instead have each module call cprover_default_properties)
but this isn't done in this commit.

Because memory analyzer depends on GDB being present and further uses platform
specific functionality at the moment it had some ifdef functionality to disable
itself.

This made the code a bit more complicated than it needed to be, and also lead to
the code effectively building defunct executables. This removes these ifdefs and
instead excludes memory-analyzer (and related tests) from the build unless
requested (via WITH_MEMORY_ANALYZER environment variable or CMake option
depending on whether it is a Make or CMake build respectively).

Also force building memory-analyzer on Linux and test it there by
default (unless explicitly unset). Behaviour on other platforms should be
preserved.

petr-bauch committed:
Add support for running regression tests under CMake

and update the chain script accordingly.

petr-bauch committed:
Improve the regression tests

mostly the description and expected output.

petr-bauch committed:
Improve building scripts

mostly white spaces and cprover_default_properties.

petr-bauch committed:
Improve gdb_api

Add memory_addresst and documentation.

petr-bauch committed:
Improve type utility functions

Better names and small code refactoring.

petr-bauch committed:
Refactor analyze_symbols

and add documentation.

petr-bauch committed:
Improve memory_analyzer parser and main

Remove ui_message_handler dependence, fix header, etc.

petr-bauch committed:
Update unit test

to the new interface.

petr-bauch committed:
Add User level documentation

to CPROVER manual.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 35abeee66386b94c8ec4874224b79e45c5fdbfc9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112065798
